core/src/main/java/de/jplag/comparison/GreedyStringTiling.java

```diff
@@ -110,39 +110,38 @@ public final JPlagComparison compare(Submission firstSubmission, Submission seco
      */
-    private JPlagComparison compareInternal(Submission leftSubmission, Submission rightSubmission) {
-        int[] leftValues = this.tokenValueMapper.getTokenValuesFor(leftSubmission);
-        int[] rightValues = this.tokenValueMapper.getTokenValuesFor(rightSubmission);
+    private JPlagComparison compareOrdered(Submission leftSubmission, Submission rightSubmission) {
+        int[] leftTokens = this.tokenValueMapper.getTokenSequenceFor(leftSubmission);
+        int[] rightTokens = this.tokenValueMapper.getTokenSequenceFor(rightSubmission);
 
-        boolean[] leftMarked = calculateInitiallyMarked(leftSubmission);
-        boolean[] rightMarked = calculateInitiallyMarked(rightSubmission);
+        boolean[] leftExcludedTokens = calculateExcludedTokens(leftSubmission);
+        boolean[] rightExcludedTokens = calculateExcludedTokens(rightSubmission);
 
-        SubsequenceHashLookupTable leftLookupTable = subsequenceHashLookupTableForSubmission(leftSubmission, leftMarked);
-        SubsequenceHashLookupTable rightLookupTable = subsequenceHashLookupTableForSubmission(rightSubmission, rightMarked);
+        RollingTokenHashTable leftLookupTable = getSubsequenceHashTableFor(leftSubmission, leftExcludedTokens);
+        RollingTokenHashTable rightLookupTable = getSubsequenceHashTableFor(rightSubmission, rightExcludedTokens);
 
         int maximumMatchLength;
         List<Match> globalMatches = new ArrayList<>();
         List<Match> ignoredMatches = new ArrayList<>();
         do {
             maximumMatchLength = minimumMatchLength;
             List<Match> iterationMatches = new ArrayList<>();
-            for (int leftStartIndex = 0; leftStartIndex < leftValues.length - maximumMatchLength; leftStartIndex++) {
-                int leftSubsequenceHash = leftLookupTable.subsequenceHashForStartIndex(leftStartIndex);
-                if (checkMark(leftMarked, leftStartIndex, leftSubmission, rightSubmission)
-                        || leftSubsequenceHash == SubsequenceHashLookupTable.NO_HASH) {
+            for (int leftStartIndex = 0; leftStartIndex < leftTokens.length - maximumMatchLength; leftStartIndex++) {
+                int leftSubsequenceHash = leftLookupTable.getHashAt(leftStartIndex);
+                if (isTokenExcludedAt(leftExcludedTokens, leftStartIndex, leftSubmission, rightSubmission)
+                        || leftSubsequenceHash == RollingTokenHashTable.NO_HASH) {
                     continue;
                 }
-                List<Integer> possiblyMatchingRightStartIndexes = rightLookupTable
-                        .startIndexesOfPossiblyMatchingSubsequencesForSubsequenceHash(leftSubsequenceHash);
-                for (Integer rightStartIndex : possiblyMatchingRightStartIndexes) {
+                List<Integer> rightStartIndices = rightLookupTable.getStartIndicesForHash(leftSubsequenceHash);
+                for (Integer rightStartIndex : rightStartIndices) { // possible matches
```

Comment:
if rightStartIndex != null -> int

Reply:
These values _currently_ cannot be null, and the check is performed a few hundred million times for realistic data sizes. Would you still include it?